scripts/pretraining/bert/pretraining_utils.py

```diff
@@ -0,0 +1,273 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Utilities for pre-training."""
+import time
+import os
+import sys
+import logging
+import random
+
+import numpy as np
+import mxnet as mx
+from mxnet.gluon.metric import EvalMetric
+
+import gluonnlp.data.batchify as bf
+from gluonnlp.utils.misc import glob
+from gluonnlp.data.loading import NumpyDataset, DatasetLoader
+from gluonnlp.data.sampler import SplitSampler, FixedBucketSampler
+
+from create_pretraining_data import create_training_instances
+
+
+def prepare_pretrain_npz_dataset(filename, allow_pickle=False):
+    """Create dataset based on the numpy npz file"""
+    if isinstance(filename, (list, tuple)):
+        assert len(filename) == 1, \
+            'When .npy/.npz data file is loaded, len(filename) must be 1.' \
+            ' Received len(filename)={}.'.format(len(filename))
+        filename = filename[0]
+    logging.debug('start to load file %s ...', filename)
+    return NumpyDataset(filename, allow_pickle=allow_pickle)
+
+
+def prepare_pretrain_text_dataset(filename, tokenizer, max_seq_length, short_seq_prob,
+                                  masked_lm_prob, max_predictions_per_seq, whole_word_mask,
+                                  vocab):
+    """Create dataset based on the raw text files"""
+    dupe_factor = 1
+    if not isinstance(filename, (list, tuple)):
+        filename = [filename]
+    logging.debug('start to load files %s ...', filename)
+    instances = create_training_instances((filename, tokenizer, max_seq_length,
+                                           short_seq_prob, masked_lm_prob,
+                                           max_predictions_per_seq,
+                                           whole_word_mask, vocab,
+                                           dupe_factor, 1, None, None))
+    return mx.gluon.data.ArrayDataset(*instances)
+
+
+def prepare_pretrain_bucket_sampler(dataset, batch_size, shuffle=False, num_buckets=1):
+    """Create data sampler based on the dataset"""
+    if isinstance(dataset, NumpyDataset):
+        lengths = dataset.get_field('valid_lengths')
+    else:
+        lengths = dataset.transform(lambda input_ids, segment_ids, masked_lm_positions, \
+                                           masked_lm_ids, masked_lm_weights, \
+                                           next_sentence_labels, valid_lengths: \
+                                        valid_lengths, lazy=False)
+    sampler = FixedBucketSampler(lengths,
+                                 batch_size=batch_size,
+                                 num_buckets=num_buckets,
+                                 ratio=0,
```

Comment:
If it's 0, is it equivalent to BatchSampler?

Reply:
Not sure. I just follow the electra and the original bert.